pkg/mapper/mapper_cache.go

```diff
@@ -0,0 +1,50 @@
+// Copyright 2019 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mapper
+
+import (
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type MetricMapperCacheResult struct {
+	Mapping *MetricMapping
+	Matched bool
+	Labels  prometheus.Labels
+}
+
+type MetricMapperCache struct {
+	cache sync.Map
```

Comment:
This cache will grow without bounds. We should probably define a maximum number of items we will cache and a cache replacement strategy,

Reply:
Any thoughts on good patterns or resources to look at for cache replacement strategies? It’s not something I am super familiar with. My original thought had been that it would be a cache with a natural limit because there should be an upper limit on the cardinality  you are pushing into the service (since the service ingesting it has a natural limit), but I think you’re right that people could be using this to take high cardinality data and reduce the cardinality before exposing it to prometheus.

My only other thought would be is it adequate to be able to opt-out of the cache on a per-metric basis? IE, if I know this is something that will have a high cardinality on the incoming event level, explicitly don’t cache it?

Reply:
I've used this in the past: https://github.com/hashicorp/golang-lru - you could have the default size be 0, which could mean no cache.

Opting in/out on a per metric basis is interesting, but I think it may complicate both the code and the user experience.

Reply:
Cool! Lemme give this a poke and see how the perf holds up. Will take ~ a week or so.

Reply:
To what extent will it grow? By design, the exporter consumes some memory per time series, so if cleaning up this cache outside of configuration reloads is difficult I would be okay with not being very clever about it until there is evidence of actual issues.

Reply:
@SpencerMalone mentioned to me that we have seen memory issues in our fork of statsd-exporter which implements this cache. We will do some testing to see if the cache is the culprit. For most people, it will probably be fine, but we have a few unusual circumstances in our environment.

Reply:
makes sense, I'm not opposed to an LRU or similar mechanism if it's worth it!